Utils/Dataflow/019_esFormat/esFormat.php

```diff
@@ -80,18 +80,46 @@ function constructActionJson($row) {
 function constructDataJson($row) {
   $data = $row;
 
+  if (isset($data['_incomplete'])) {
+    $incompl = $data['_incomplete'];
+  }
+
   foreach ($data as $key => $val) {
     if (strncmp($key, '_', 1) === 0)
       unset($data[$key]);
   }
 
   $act = getAction($row);
+  $insert_data = $data;
+  $update_data = $data;
+  if (isset($incompl)) {
+    # We *may* not specify "_update_required" value on insert, if it is False,
+    # (for it is a default value), but to allow usage of 'doc_as_upsert'
+    # ES option, it can be specified explicitly. So in case of 'update' action
+    # it should always be so.
```

Comment:
This comment is confusing. I've read d44ec42's description and the code block, but still have hard time understanding the comment. How about:

    There are four possible situations depending on the combination
    of two values: action (insert or update) and data incompleteness = "_update_required" (True or False).
    While there is no need to explicitly set "_update_required" value to False
    (for it is a default value), doing so in case of update will allow usage of 'doc_as_upsert'
    (update that adds new, complete document that was missing).
    This means that in three out of four cases - insert of incomplete data
    and update of data of any kind - the value must be set explicitly.

Correct me if I'm wrong.
P.S. Suggested comment, if accepted, should be checked for line length.

Reply:
@Evildoor 
> ```
> ...usage of 'doc_as_upsert'
> (update that adds new, complete document that was missing).
> ```

1. In fact, `doc_as_upsert` allows us to send one document (for update of an existing one and to be inserted if document does not exist) instead of two (one for update, another -- for insert), so I would say that the suggested version is incorrect or at least inaccurate.

2. I don\`t think we need to talk here about all four scenarios we have: this comment is related only to the statement right after it, which says "set the `_update_required` field explicitly, if action is 'update' or message is incomplete".

3. I still will rewrite this part, since it is obviously not very clear :)

Reply:
Aha, I see. New variant looks fine to me, apart from some language-related nuances (commented below).